crates/bevy_render/src/extract_param.rs

```diff
@@ -0,0 +1,120 @@
+use crate::MainWorld;
+use bevy_ecs::{
+    prelude::*,
+    system::{
+        ReadOnlySystemParamFetch, ResState, SystemMeta, SystemParam, SystemParamFetch,
+        SystemParamState, SystemState,
+    },
+};
+use std::ops::{Deref, DerefMut};
+
+/// A helper for accessing [`MainWorld`] content using a system parameter.
+///
+/// A [`SystemParam`] adapter which applies the contained `SystemParam` to the [`World`]
+/// contained in [`MainWorld`]. This parameter only works for systems run
+/// during [`RenderStage::Extract`].
+///
+/// This requires that the contained [`SystemParam`] does not mutate the world, as it
+/// uses a read-only reference to [`MainWorld`] internally.
+///
+/// ## Context
+///
+/// [`RenderStage::Extract`] is used to extract (move) data from the simulation world ([`MainWorld`]) to the
+/// render world. The render world drives rendering each frame (generally to a [Window]).
+/// This design is used to allow performing calculations related to rendering a prior frame at the same
+/// time as the next frame is simulated, which increases throughput (FPS).
+///
+/// [`Extract`] is used to get data from the main world during [`RenderStage::Extract`].
+///
+/// ## Examples
+///
+/// ```rust
+/// use bevy_ecs::prelude::*;
+/// use bevy_render::Extract;
+/// # #[derive(Component)]
+/// # struct Cloud;
+/// fn extract_clouds(mut commands: Commands, clouds: Extract<Query<Entity, With<Cloud>>>) {
+///     for cloud in clouds.iter() {
+///         commands.get_or_spawn(cloud).insert(Cloud);
+///     }
+/// }
+/// ```
+///
+/// [`RenderStage::Extract`]: crate::RenderStage::Extract
+/// [Window]: bevy_window::Window
+pub struct Extract<'w, 's, P: SystemParam + 'static>
+where
+    P::Fetch: ReadOnlySystemParamFetch,
+{
+    item: <P::Fetch as SystemParamFetch<'w, 's>>::Item,
```

Comment:
When I tried this previously (https://github.com/bevyengine/bevy/commit/6640933d981ed990ac20586368cce3d1ad032364), I got into a muddle about these lifetimes; I believed that I needed a lifetime which referred to the `main_world` in `get_param`.

However, I now see that since all the references are passed down with the same lifetime, it all works out.